va/va.go

```diff
@@ -442,110 +442,106 @@ func (va *ValidationAuthorityImpl) validateChallenge(
-// channel as-is.
+// performRemoteValidation coordinates the whole process of kicking off and
+// collecting results from calls to remote VAs' PerformValidation function. It
+// returns either a problem (if too many of the RVAs observed validation
+// failures), or an error (if too many of the remote validation requests
+// encountered non-validation errors), or neither if the number of successful
+// remote validations surpassed our corroboration threshold.
 func (va *ValidationAuthorityImpl) performRemoteValidation(
 	ctx context.Context,
 	req *vapb.PerformValidationRequest,
-	results chan<- *remoteVAResult) {
-	for _, i := range rand.Perm(len(va.remoteVAs)) {
-		remoteVA := va.remoteVAs[i]
-		go func(rva RemoteVA) {
-			result := &remoteVAResult{
-				VAHostname: rva.Address,
-			}
-			res, err := rva.PerformValidation(ctx, req)
-			if err != nil && canceled.Is(err) {
-				// If the non-nil err was a canceled error, ignore it. That's fine: it
-				// just means we cancelled the remote VA request before it was
-				// finished because we didn't care about its result. Don't log to avoid
-				// spamming the logs.
-				result.Problem = probs.ServerInternal("Remote PerformValidation RPC canceled")
-			} else if err != nil {
-				// This is a real error, not just a problem with the validation.
-				va.log.Errf("Remote VA %q.PerformValidation failed: %s", rva.Address, err)
-				result.Problem = probs.ServerInternal("Remote PerformValidation RPC failed")
-			} else if res.Problems != nil {
-				prob, err := bgrpc.PBToProblemDetails(res.Problems)
-				if err != nil {
-					va.log.Infof("Remote VA %q.PerformValidation returned malformed problem: %s", rva.Address, err)
-					result.Problem = probs.ServerInternal(
-						fmt.Sprintf("Remote PerformValidation RPC returned malformed result: %s", err))
-				} else {
-					va.log.Infof("Remote VA %q.PerformValidation returned problem: %s", rva.Address, prob)
-					result.Problem = prob
-				}
-			}
-			results <- result
-		}(remoteVA)
+) (prob *probs.ProblemDetails) {
+	if len(va.remoteVAs) == 0 {
+		return nil
 	}
-}
-
-// processRemoteValidationResults evaluates a primary VA result, and a channel
-// of remote VA problems to produce a single overall validation result based on
-// configured feature flags. The overall result is calculated based on the VA's
-// configured `maxRemoteFailures` value, and the function returns as soon as
-// that threshold has been exceeded or cannot possibly be exceeded.
-func (va *ValidationAuthorityImpl) processRemoteValidationResults(
-	challengeType string,
-	remoteResultsChan <-chan *remoteVAResult) *probs.ProblemDetails {
 
-	state := "failure"
+	// Defer a closure to increment a metric. Note that this function uses named
+	// return values so that this closure can reference them.
 	start := va.clk.Now()
-
 	defer func() {
+		state := "failure"
+		if prob == nil {
+			state = "success"
+		}
```

Comment:
I think this function is a particularly bad candidate for the "named returns + deferred func" technique. Returning a `nil` value means "success", which we should be careful about doing; we only want to do it when we've reached our threshold of good responses.

But making `prob` a named return value, its initial value is `nil`. That means a bare `return` could return `nil` (i.e. success), but that wouldn't be obvious from reading that line of code (which would be far from the named return value).

Instead, let's just remove `result` from the `remoteValidationTime` metric. I don't think it's useful, and that lets us easily skip the defer trick.

Reply:
Good call, it doesn't look like any of the graphs on the RVA dashboard split by `result`.